// ng-cdk-table-virtual-scroll/bundles/ng-cdk-table-virtual-scroll.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/operators'), require('rxjs'), require('@angular/cdk/scrolling'), require('@angular/cdk/table')) :
    typeof define === 'function' && define.amd ? define('ng-cdk-table-virtual-scroll', ['exports', '@angular/core', 'rxjs/operators', 'rxjs', '@angular/cdk/scrolling', '@angular/cdk/table'], factory) :
    (global = global || self, factory(global['ng-cdk-table-virtual-scroll'] = {}, global.ng.core, global.rxjs.operators, global.rxjs, global.ng.cdk.scrolling, global.ng.cdk.table));
}(this, (function (exports, core, operators, rxjs, scrolling, table) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var TableVirtualScrollDataSource = /** @class */ (function (_super) {
        __extends(TableVirtualScrollDataSource, _super);
        function TableVirtualScrollDataSource(initialData) {
            if (initialData === void 0) { initialData = []; }
            var _this = _super.call(this) || this;
            /** Stream emitting render data to the table (depends on ordered data changes). */
            _this._renderData = new rxjs.BehaviorSubject([]);
            /**
             * Subscription to the changes that should trigger an update to the table's rendered rows, such
             * as filtering, sorting, pagination, or base data changes.
             */
            _this._renderChangesSubscription = null;
            _this._data = new rxjs.BehaviorSubject(initialData);
            _this.updateChangeSubscription();
            return _this;
        }
        Object.defineProperty(TableVirtualScrollDataSource.prototype, "data", {
            /** Array of data that should be rendered by the table, where each object represents one row. */
            get: function () {
                return this._data.value;
            },
            set: function (data) {
                this._data.next(data);
            },
            enumerable: true,
            configurable: true
        });
        TableVirtualScrollDataSource.prototype.updateChangeSubscription = function () {
            var _this = this;
            this.initStreams();
            this._renderChangesSubscription = new rxjs.Subscription();
            this._renderChangesSubscription.add(this._data.subscribe(function (data) { return _this.dataToRender$.next(data); }));
            this._renderChangesSubscription.add(this.dataOfRange$.subscribe(function (data) { return _this._renderData.next(data); }));
        };
        TableVirtualScrollDataSource.prototype.initStreams = function () {
            if (!this.streamsReady) {
                this.dataToRender$ = new rxjs.ReplaySubject(1);
                this.dataOfRange$ = new rxjs.ReplaySubject(1);
                this.streamsReady = true;
            }
        };
        /**
         * Used by the MatTable. Called when it connects to the data source.
         * @private
         */
        TableVirtualScrollDataSource.prototype.connect = function () {
            if (!this._renderChangesSubscription) {
                this.updateChangeSubscription();
            }
            return this._renderData;
        };
        /**
         * Used by the MatTable. Called when it disconnects from the data source.
         * @private
         */
        TableVirtualScrollDataSource.prototype.disconnect = function () {
            if (this._renderChangesSubscription) {
                this._renderChangesSubscription.unsubscribe();
            }
            this._renderChangesSubscription = null;
        };
        return TableVirtualScrollDataSource;
    }(table.DataSource));

    var FixedSizeTableVirtualScrollStrategy = /** @class */ (function () {
        function FixedSizeTableVirtualScrollStrategy() {
            this.indexChange = new rxjs.Subject();
            this.stickyChange = new rxjs.Subject();
            this.renderedRangeStream = new rxjs.BehaviorSubject({ start: 0, end: 0 });
            this.scrolledIndexChange = this.indexChange.pipe(operators.distinctUntilChanged());
            this._dataLength = 0;
        }
        Object.defineProperty(FixedSizeTableVirtualScrollStrategy.prototype, "dataLength", {
            get: function () {
                return this._dataLength;
            },
            set: function (value) {
                this._dataLength = value;
                this.onDataLengthChanged();
            },
            enumerable: true,
            configurable: true
        });
        FixedSizeTableVirtualScrollStrategy.prototype.attach = function (viewport) {
            this.viewport = viewport;
            this.viewport.renderedRangeStream.subscribe(this.renderedRangeStream);
            this.onDataLengthChanged();
        };
        FixedSizeTableVirtualScrollStrategy.prototype.detach = function () {
            this.indexChange.complete();
            this.stickyChange.complete();
            this.renderedRangeStream.complete();
        };
        FixedSizeTableVirtualScrollStrategy.prototype.onContentScrolled = function () {
            this.updateContent();
        };
        FixedSizeTableVirtualScrollStrategy.prototype.onDataLengthChanged = function () {
            if (this.viewport) {
                this.viewport.setTotalContentSize(this.dataLength * this.rowHeight + this.headerHeight + this.footerHeight);
            }
            this.updateContent();
        };
        FixedSizeTableVirtualScrollStrategy.prototype.onContentRendered = function () {
            // no-op
        };
        FixedSizeTableVirtualScrollStrategy.prototype.onRenderedOffsetChanged = function () {
            // no-op
        };
        FixedSizeTableVirtualScrollStrategy.prototype.scrollToIndex = function (index, behavior) {
            if (!this.viewport || !this.rowHeight) {
                return;
            }
            this.viewport.scrollToOffset((index - 1) * this.rowHeight + this.headerHeight);
        };
        FixedSizeTableVirtualScrollStrategy.prototype.setConfig = function (configs) {
            var rowHeight = configs.rowHeight, headerHeight = configs.headerHeight, footerHeight = configs.footerHeight, bufferMultiplier = configs.bufferMultiplier;
            if (this.rowHeight === rowHeight
                && this.headerHeight === headerHeight
                && this.footerHeight === footerHeight
                && this.bufferMultiplier === bufferMultiplier) {
                return;
            }
            this.rowHeight = rowHeight;
            this.headerHeight = headerHeight;
            this.footerHeight = footerHeight;
            this.bufferMultiplier = bufferMultiplier;
            this.onDataLengthChanged();
        };
        FixedSizeTableVirtualScrollStrategy.prototype.updateContent = function () {
            if (!this.viewport || !this.rowHeight) {
                return;
            }
            var scrollOffset = this.viewport.measureScrollOffset();
            var amount = Math.ceil(this.viewport.getViewportSize() / this.rowHeight);
            var offset = Math.max(scrollOffset - this.headerHeight, 0);
            var buffer = Math.ceil(amount * this.bufferMultiplier);
            var skip = Math.round(offset / this.rowHeight);
            var index = Math.max(0, skip);
            var start = Math.max(0, index - buffer);
            var end = Math.min(this.dataLength, index + amount + buffer);
            var renderedOffset = start * this.rowHeight;
            this.viewport.setRenderedContentOffset(renderedOffset);
            this.viewport.setRenderedRange({ start: start, end: end });
            this.indexChange.next(index);
            this.stickyChange.next(renderedOffset);
        };
        FixedSizeTableVirtualScrollStrategy = __decorate([
            core.Injectable()
        ], FixedSizeTableVirtualScrollStrategy);
        return FixedSizeTableVirtualScrollStrategy;
    }());

    function _tableVirtualScrollDirectiveStrategyFactory(tableDir) {
        return tableDir.scrollStrategy;
    }
    var stickyHeaderSelector = '.cdk-header-row .cdk-table-sticky';
    var stickyFooterSelector = '.cdk-footer-row .cdk-table-sticky';
    var defaults = {
        rowHeight: 48,
        headerHeight: 56,
        headerEnabled: true,
        footerHeight: 48,
        footerEnabled: false,
        bufferMultiplier: 0.7,
    };
    var TableItemSizeDirective = /** @class */ (function () {
        function TableItemSizeDirective(zone) {
            this.zone = zone;
            this.alive = true;
            // tslint:disable-next-line:no-input-rename
            this.rowHeight = defaults.rowHeight;
            this.headerEnabled = defaults.headerEnabled;
            this.headerHeight = defaults.headerHeight;
            this.footerEnabled = defaults.footerEnabled;
            this.footerHeight = defaults.footerHeight;
            this.bufferMultiplier = defaults.bufferMultiplier;
            this.scrollStrategy = new FixedSizeTableVirtualScrollStrategy();
            this.dataSourceChanges = new rxjs.Subject();
        }
        TableItemSizeDirective_1 = TableItemSizeDirective;
        TableItemSizeDirective.prototype.ngOnDestroy = function () {
            this.alive = false;
            this.dataSourceChanges.complete();
        };
        TableItemSizeDirective.prototype.isAlive = function () {
            var _this = this;
            return function () { return _this.alive; };
        };
        TableItemSizeDirective.prototype.isStickyEnabled = function () {
            return (!!this.scrollStrategy.viewport &&
                this.table['_headerRowDefs']
                    .map(function (def) { return def.sticky; })
                    .reduce(function (prevState, state) { return prevState && state; }, true));
        };
        TableItemSizeDirective.prototype.ngAfterContentInit = function () {
            var _this = this;
            var switchDataSourceOrigin = this.table['_switchDataSource'];
            this.table['_switchDataSource'] = function (dataSource) {
                switchDataSourceOrigin.call(_this.table, dataSource);
                _this.connectDataSource(dataSource);
            };
            this.connectDataSource(this.table.dataSource);
            this.scrollStrategy.stickyChange
                .pipe(operators.filter(function () { return _this.isStickyEnabled(); }), operators.tap(function () {
                if (!_this.stickyPositions) {
                    _this.initStickyPositions();
                }
            }), operators.takeWhile(this.isAlive()))
                .subscribe(function (stickyOffset) {
                _this.setSticky(stickyOffset);
            });
        };
        TableItemSizeDirective.prototype.connectDataSource = function (dataSource) {
            var _this = this;
            this.dataSourceChanges.next();
            if (dataSource instanceof TableVirtualScrollDataSource) {
                dataSource.dataToRender$
                    .pipe(operators.distinctUntilChanged(), operators.takeUntil(this.dataSourceChanges), operators.takeWhile(this.isAlive()), operators.tap(function (data) { return (_this.scrollStrategy.dataLength = data.length); }), operators.switchMap(function (data) {
                    return _this.scrollStrategy.renderedRangeStream.pipe(operators.map(function (_a) {
                        var start = _a.start, end = _a.end;
                        return typeof start !== 'number' || typeof end !== 'number'
                            ? data
                            : data.slice(start, end);
                    }));
                }))
                    .subscribe(function (data) {
                    _this.zone.run(function () {
                        dataSource.dataOfRange$.next(data);
                    });
                });
            }
            else {
                throw new Error('[tvsItemSize] requires TableVirtualScrollDataSource be set as [dataSource] of [cdk-table]');
            }
        };
        TableItemSizeDirective.prototype.ngOnChanges = function () {
            var config = {
                rowHeight: +this.rowHeight || defaults.rowHeight,
                headerHeight: this.headerEnabled
                    ? +this.headerHeight || defaults.headerHeight
                    : 0,
                footerHeight: this.footerEnabled
                    ? +this.footerHeight || defaults.footerHeight
                    : 0,
                bufferMultiplier: +this.bufferMultiplier || defaults.bufferMultiplier,
            };
            this.scrollStrategy.setConfig(config);
        };
        TableItemSizeDirective.prototype.setSticky = function (offset) {
            var _this = this;
            this.scrollStrategy.viewport.elementRef.nativeElement
                .querySelectorAll(stickyHeaderSelector)
                .forEach(function (el) {
                var parent = el.parentElement;
                var baseOffset = 0;
                if (_this.stickyPositions.has(parent)) {
                    baseOffset = _this.stickyPositions.get(parent);
                }
                el.style.top = baseOffset - offset + "px";
            });
            this.scrollStrategy.viewport.elementRef.nativeElement
                .querySelectorAll(stickyFooterSelector)
                .forEach(function (el) {
                var parent = el.parentElement;
                var baseOffset = 0;
                if (_this.stickyPositions.has(parent)) {
                    baseOffset = _this.stickyPositions.get(parent);
                }
                el.style.bottom = -baseOffset + offset + "px";
            });
        };
        TableItemSizeDirective.prototype.initStickyPositions = function () {
            var _this = this;
            this.stickyPositions = new Map();
            this.scrollStrategy.viewport.elementRef.nativeElement
                .querySelectorAll(stickyHeaderSelector)
                .forEach(function (el) {
                var parent = el.parentElement;
                if (!_this.stickyPositions.has(parent)) {
                    _this.stickyPositions.set(parent, parent.offsetTop);
                }
            });
        };
        var TableItemSizeDirective_1;
        TableItemSizeDirective.ctorParameters = function () { return [
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input('tvsItemSize')
        ], TableItemSizeDirective.prototype, "rowHeight", void 0);
        __decorate([
            core.Input()
        ], TableItemSizeDirective.prototype, "headerEnabled", void 0);
        __decorate([
            core.Input()
        ], TableItemSizeDirective.prototype, "headerHeight", void 0);
        __decorate([
            core.Input()
        ], TableItemSizeDirective.prototype, "footerEnabled", void 0);
        __decorate([
            core.Input()
        ], TableItemSizeDirective.prototype, "footerHeight", void 0);
        __decorate([
            core.Input()
        ], TableItemSizeDirective.prototype, "bufferMultiplier", void 0);
        __decorate([
            core.ContentChild(table.CdkTable, { static: false })
        ], TableItemSizeDirective.prototype, "table", void 0);
        TableItemSizeDirective = TableItemSizeDirective_1 = __decorate([
            core.Directive({
                selector: 'cdk-virtual-scroll-viewport[tvsItemSize]',
                providers: [
                    {
                        provide: scrolling.VIRTUAL_SCROLL_STRATEGY,
                        useFactory: _tableVirtualScrollDirectiveStrategyFactory,
                        deps: [core.forwardRef(function () { return TableItemSizeDirective_1; })],
                    },
                ],
            })
        ], TableItemSizeDirective);
        return TableItemSizeDirective;
    }());

    var TableVirtualScrollModule = /** @class */ (function () {
        function TableVirtualScrollModule() {
        }
        TableVirtualScrollModule = __decorate([
            core.NgModule({
                declarations: [TableItemSizeDirective],
                exports: [TableItemSizeDirective],
                imports: [],
            })
        ], TableVirtualScrollModule);
        return TableVirtualScrollModule;
    }());

    exports.FixedSizeTableVirtualScrollStrategy = FixedSizeTableVirtualScrollStrategy;
    exports.TableItemSizeDirective = TableItemSizeDirective;
    exports.TableVirtualScrollDataSource = TableVirtualScrollDataSource;
    exports.TableVirtualScrollModule = TableVirtualScrollModule;
    exports._tableVirtualScrollDirectiveStrategyFactory = _tableVirtualScrollDirectiveStrategyFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
