// ng-cdk-table-virtual-scroll/esm5/lib/table-data-source.js
import { __extends } from "tslib";
import { DataSource } from '@angular/cdk/table';
import { BehaviorSubject, ReplaySubject, Subscription, } from 'rxjs';
var TableVirtualScrollDataSource = /** @class */ (function (_super) {
    __extends(TableVirtualScrollDataSource, _super);
    function TableVirtualScrollDataSource(initialData) {
        if (initialData === void 0) { initialData = []; }
        var _this = _super.call(this) || this;
        /** Stream emitting render data to the table (depends on ordered data changes). */
        _this._renderData = new BehaviorSubject([]);
        /**
         * Subscription to the changes that should trigger an update to the table's rendered rows, such
         * as filtering, sorting, pagination, or base data changes.
         */
        _this._renderChangesSubscription = null;
        _this._data = new BehaviorSubject(initialData);
        _this.updateChangeSubscription();
        return _this;
    }
    Object.defineProperty(TableVirtualScrollDataSource.prototype, "data", {
        /** Array of data that should be rendered by the table, where each object represents one row. */
        get: function () {
            return this._data.value;
        },
        set: function (data) {
            this._data.next(data);
        },
        enumerable: true,
        configurable: true
    });
    TableVirtualScrollDataSource.prototype.updateChangeSubscription = function () {
        var _this = this;
        this.initStreams();
        this._renderChangesSubscription = new Subscription();
        this._renderChangesSubscription.add(this._data.subscribe(function (data) { return _this.dataToRender$.next(data); }));
        this._renderChangesSubscription.add(this.dataOfRange$.subscribe(function (data) { return _this._renderData.next(data); }));
    };
    TableVirtualScrollDataSource.prototype.initStreams = function () {
        if (!this.streamsReady) {
            this.dataToRender$ = new ReplaySubject(1);
            this.dataOfRange$ = new ReplaySubject(1);
            this.streamsReady = true;
        }
    };
    /**
     * Used by the MatTable. Called when it connects to the data source.
     * @private
     */
    TableVirtualScrollDataSource.prototype.connect = function () {
        if (!this._renderChangesSubscription) {
            this.updateChangeSubscription();
        }
        return this._renderData;
    };
    /**
     * Used by the MatTable. Called when it disconnects from the data source.
     * @private
     */
    TableVirtualScrollDataSource.prototype.disconnect = function () {
        if (this._renderChangesSubscription) {
            this._renderChangesSubscription.unsubscribe();
        }
        this._renderChangesSubscription = null;
    };
    return TableVirtualScrollDataSource;
}(DataSource));
export { TableVirtualScrollDataSource };
